### ISSUE TITLE
Backpropagation fails with CUDA_ERROR_ILLEGAL_ADDRESS

### ISSUE THREAD

**Issue by rm-dr** [tracel-ai/burn#4187]
## Describe the bug:
I'm encountering a CUDA_ERROR_ILLEGAL_ADDRESS while running backpropagation. I tested this code with the `wgpu` and `ndarray` backends, they seem to work. `cuda` does not. 

- Stack trace (`RUST_BACKTRACE=full`): [stacktrace.txt](https://github.com/user-attachments/files/24154628/stacktrace.txt)
- `compute-sanitizer` output of failing binary: [cs.txt](https://github.com/user-attachments/files/24154632/cs.txt)

Entire code is [here](https://git.betalupi.com/Mark/llmfs). Might not be useful, there is a lot. Might make a MWE later.

## Dependencies:
- Rust version is 1.91.1.
- OS is linux
- CUDA is 13.0
- Burn deps are `0.19.1`, but this is easily reproduced on main (https://github.com/tracel-ai/burn/commit/ddb8de68aec77b792fdd365b29896d1939403fd6)
```toml
burn-train = { version = "0.19.1", default-features = false }

[workspace.dependencies.burn]
version = "0.19.1"
default-features = false
features = ["std", "fusion", "ndarray", "webgpu", "cuda", "autodiff"]
```

## Additional context:
First mentioned in discord: https://discord.com/channels/1038839012602941528/1449845465842454675

**Comment by rm-dr** [comment]
# MWE

## Cargo.toml:
```toml
[package]
name = "burnbug"
version = "0.1.0"
edition = "2024"

[dependencies]
burn-train = { version = "0.19.1", default-features = false }

[dependencies.burn]
version = "0.19.1"
default-features = false
features = ["std", "fusion", "ndarray", "webgpu", "cuda", "autodiff"]
```


## main.rs:

<details>


```rust
use burn::{
    Tensor,
    backend::{Autodiff, Cuda, cuda::CudaDevice},
    config::Config,
    module::{Module, Param, ParamId},
    nn::{
        Dropout, Embedding, EmbeddingConfig, LayerNorm, LayerNormConfig,
        loss::CrossEntropyLossConfig,
    },
    optim::{AdamConfig, GradientsParams, Optimizer},
    prelude::Backend,
    tensor::{Distribution, Int},
};
use burn_train::ClassificationOutput;

pub fn main() {
    type Backend = Cuda;
    let device = CudaDevice::new(0);

    let mcfg = GptModelConfig {
        vocab_size: 10,
        context_size: 256,
        embed_dim: 768,
        n_heads: 12,
        head_dim: 64,
        embed_drop_rate: 0.1,
        attention_drop_rate: 0.1,
        shortcut_drop_rate: 0.1,
    };

    let mut model: GptModel<Autodiff<Backend>> = mcfg.init(&device);

    let mut optim = AdamConfig::new().init();
    let learning_rate = 1e-4;

    for epoch in 0..10 {
        for batch in (0..10).map(|_| {
            (
                // training data goes here
                Tensor::<_, 2, Int>::random(
                    [10, mcfg.context_size],
                    Distribution::Default,
                    &device,
                ),
                Tensor::<_, 1, Int>::random([10], Distribution::Default, &device),
            )
        }) {
            let (inputs, targets) = batch;
            let output = model.forward_train(inputs, targets);

            println!("Running backward pass");
            let grads = output.loss.backward();
            let grads = GradientsParams::from_grads(grads, &model);

            println!("Running optimizer step");
            model = optim.step(learning_rate, model, grads);
        }

        println!("Ran epoch {epoch}");
    }
}

//
// MARK: model
//

#[derive(Config, Debug)]
pub struct GptModelConfig {
    /// Number of tokens
    pub vocab_size: u32,

    /// Maximum number of input tokens with positional embeddings
    pub context_size: usize,

    /// Dimension of each token's embedding
    pub embed_dim: usize,

    /// Number of attention heads
    pub n_heads: usize,

    /// Dimension of each attn head
    pub head_dim: usize,

    pub embed_drop_rate: f64,
    pub attention_drop_rate: f64,
    pub shortcut_drop_rate: f64,
}

impl GptModelConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> GptModel<B> {
        let out_head_shape = [self.embed_dim, self.vocab_size as usize];

        GptModel {
            embedder_tok: EmbeddingConfig::new(self.vocab_size as usize, self.embed_dim)
                .init(device),

            embedder_pos: EmbeddingConfig::new(self.context_size, self.embed_dim).init(device),

            embedder_drop: Dropout {
                prob: self.embed_drop_rate,
            },

            final_norm: LayerNormConfig::new(self.embed_dim).init(device),

            out_head: Param::uninitialized(
                ParamId::new(),
                move |device, is_require_grad| {
                    Tensor::random(out_head_shape, Distribution::Default, device)
                        .set_require_grad(is_require_grad)
                },
                device.clone(),
                true,
                out_head_shape.into(),
            ),
        }
    }
}

#[derive(Debug, Clone)]
pub struct TrainBatch<B: Backend> {
    pub inputs: Tensor<B, 2, Int>,

    /// Correct next token for each input
    pub targets: Tensor<B, 1, Int>,
}

#[derive(Module, Debug)]
pub struct GptModel<B: Backend> {
    embedder_tok: Embedding<B>,
    embedder_pos: Embedding<B>,
    embedder_drop: Dropout,

    final_norm: LayerNorm<B>,
    out_head: Param<Tensor<B, 2>>,
}

impl<B: Backend> GptModel<B> {
    pub fn forward(&self, input: Tensor<B, 2, Int>) -> Tensor<B, 3> {
        let n_tokens = input.shape()[1];

        let embed_tok = self.embedder_tok.forward(input.clone());
        let embed_pos = self
            .embedder_pos
            .forward(Tensor::arange(0..n_tokens as i64, &input.device()).unsqueeze_dim(0));

        let x = embed_tok + embed_pos;
        let x = self.embedder_drop.forward(x);
        let x = self.final_norm.forward(x);
        let logits = x.matmul(self.out_head.val().unsqueeze_dim(0));

        return logits;
    }

    pub fn forward_train(
        &self,
        inputs: Tensor<B, 2, Int>,
        targets: Tensor<B, 1, Int>,
    ) -> ClassificationOutput<B> {
        // shape: [batch, n_tokens, n_vocabulary]
        let output = self.forward(inputs);

        // Get last token
        // shape: [batch, n_vocabulary]
        let output = output.slice_dim(1, -1).squeeze_dim::<2>(1);

        let loss = CrossEntropyLossConfig::new()
            .init(&targets.device())
            .forward(output.clone(), targets.clone());

        ClassificationOutput {
            loss,
            output,
            targets,
        }
    }
}
```
</details>

**Comment by laggui** [comment]
Thanks for filing the issue with a MWE!

Just had a look, using `CUBECL_DEBUG_LOG` (forces sync for each execution, and we can inspect the source):
```
Running backward pass

thread 'main' (858006) panicked at /home/laggui/workspace/burn/crates/burn-cubecl/src/kernel/index/slice_assign.rs:12:1:
called `Result::unwrap()` on an `Err` value: An execution error happened during profiling
Caused by:
  An error happened during execution
Caused by:
  DriverError(CUDA_ERROR_ILLEGAL_ADDRESS, "an illegal memory access was encountered")
Backtrace:
No backtrace available
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: cubecl_runtime::client::ComputeClient<R>::launch_inner
   4: cubecl_runtime::client::ComputeClient<R>::launch_unchecked
   5: cubecl_core::compute::launcher::KernelLauncher<R>::launch_unchecked
   6: burn_cubecl::kernel::index::slice_assign::slice_assign_kernel::launch_unchecked
   7: burn_cubecl::kernel::index::slice_assign::slice_assign
   8: <burn_autodiff::ops::tensor::<impl burn_backend::backend::ops::tensor::FloatTensorOps<burn_autodiff::backend::Autodiff<B,C>> for burn_autodiff::backend::Autodiff<B,C>>::float_slice::Index as burn_autodiff::ops::backward::Backward<B,1_usize>>::backward
   9: <burn_autodiff::ops::base::OpsStep<B,T,SB,_> as burn_autodiff::graph::base::Step>::step
  10: burn_autodiff::runtime::server::AutodiffServer::execute_steps
  11: gh_4004::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

So it is caused by the `slice_assign` kernel, in this case triggered by the `tensor.slice(...)` backward pass.

**Comment by laggui** [comment]
Turns out this is related to what I was fixing in https://github.com/tracel-ai/burn/pull/4152

Didn't realize it didn't get merged 😅 

The slices in the backward kernel were not normalized (due to the high-level `slice_dim` not handling this, and leaving backends to handle this explicitly).

**Comment by rm-dr** [comment]
Just tested this on latest main, it seems to work!

**Comment by laggui** [comment]
Thanks for confirming!
